grabbit/src/main/groovy/com/twcable/grabbit/server/services/NodeWithMandatoryIterator.groovy

```diff
@@ -0,0 +1,90 @@
+/*
+ * Copyright 2015 Time Warner Cable, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.twcable.grabbit.server.services
+
+import com.twcable.grabbit.jcr.JCRNodeDecorator
+import groovy.transform.CompileStatic
+import groovy.transform.TailRecursive
+
+import javax.jcr.Node as JcrNode
+
+/**
+ * Custom Node Iterator that will iterate over a rootNode node and its immediate children
+ * taking into account cases where a node (rootNode node or any of the children) has any number of
+ * mandatory subnodes
+ */
+@CompileStatic
+final class NodeWithMandatoryIterator implements Iterator<JcrNode> {
+
+    private boolean doneRoot
+    private JCRNodeDecorator rootNode
+    private Iterator<JCRNodeDecorator> immediateChildren
+    private Iterator<JCRNodeDecorator> mandatoryWriteNodes
+
+
+    public NodeWithMandatoryIterator(JcrNode root) {
+        this.rootNode = new JCRNodeDecorator(root)
+        this.doneRoot = false
+        //Get all immediate children that are not mandatory write nodes.  We will handle those by iterating over mandatoryWriteNodes
+        immediateChildren = rootNode.getImmediateChildNodes().findAll { !it.isRequiredNode() }.iterator() as Iterator<JCRNodeDecorator>
+        //Calls a tail recursive method, gathering all required nodes
+        mandatoryWriteNodes = addToMandatoryIterator(rootNode, []).iterator()
```

Comment:
Well, in the case of a configuration path "foo/somepage/." , you're saying you just want to grab the page + page properties.  I don't think it makes sense to create compexlity, and surprise results for a few node types;  Just let the node definition define what is important.

If one wanted everything from a single page, why not explicitly express what you want;  i.e do the regular recursive syntax, and exclude subpages you don't want?